core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSource.java

```diff
@@ -119,180 +109,88 @@
         this.queryId = requireNonNull(queryId, "queryId is null");
         this.tableExecuteContextManager = requireNonNull(tableExecuteContextManager, "tableExecuteContextManager is null");
         this.sourceExchanges = ImmutableMap.copyOf(requireNonNull(sourceExchanges, "sourceExchanges is null"));
-        this.remoteSources = ImmutableMap.copyOf(requireNonNull(remoteSources, "remoteSources is null"));
-        checkArgument(
-                sourceExchanges.keySet().equals(remoteSources.keySet()),
-                "sourceExchanges and remoteSources are expected to contain the same set of keys: %s != %s",
-                sourceExchanges.keySet(),
-                remoteSources.keySet());
+        this.remoteSources = ImmutableSetMultimap.copyOf(requireNonNull(remoteSources, "remoteSources is null"));
         this.splitSourceSupplier = requireNonNull(splitSourceSupplier, "splitSourceSupplier is null");
         this.assigner = requireNonNull(assigner, "assigner is null");
-        this.callback = requireNonNull(callback, "callback is null");
         this.executor = requireNonNull(executor, "executor is null");
         this.splitBatchSize = splitBatchSize;
         this.targetExchangeSplitSizeInBytes = targetExchangeSplitSizeInBytes;
         this.sourcePartitioningScheme = requireNonNull(sourcePartitioningScheme, "sourcePartitioningScheme is null");
         this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
-        remoteSourceFragments = remoteSources.entrySet().stream()
-                .collect(toImmutableSetMultimap(Map.Entry::getValue, Map.Entry::getKey));
     }
 
-    public synchronized void start()
+    public synchronized ListenableFuture<AssignmentResult> process()
```

Comment:
If this method is called twice, and second call is done before future returned by the first call is done then we would register same splits twice in the assigner. I think we should guard against it.

Reply:
Maybe you just need to  verify that `getSplitBatchAndAdvance` is never called twice.

Reply:
Added a check to make sure `process` is not called before the previous one is finished